File: mvc-arch/src/main/java/org/example/mvc/repository/JdbcTemplate.java
package org.example.mvc.repository;

import org.example.mvc.modelAndView.view.model.User;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class JdbcTemplate {
    public void executeUpdate(String sql, PreparedStatementSetter pss) throws SQLException {
        try(
            Connection conn = ConnectionManager.getConnection();
            PreparedStatement pstmt = conn.prepareStatement(sql)
        ) {
            pss.setter(pstmt);
            pstmt.executeUpdate();
        }
    }

    public List<User> executeQuery(String sql) throws SQLException {
        List<User> findUsers = new ArrayList<>();

        try(
            Connection conn = ConnectionManager.getConnection();
            PreparedStatement pstmt = conn.prepareStatement(sql)
        ) {
            ResultSet resultSet = pstmt.executeQuery();

            while (resultSet.next()) {
                String userId = resultSet.getString("userId");
                String name = resultSet.getString("name");
                findUsers.add(new User(userId, name));
            }
        }

        return findUsers;
    }
}

File: FastCampus/jdbc/src/main/kotlin/org/example/JdbcTemplate.kt
package org.example

import java.sql.Connection
import java.sql.PreparedStatement

class JdbcTemplate {
    private lateinit var conn: Connection
    private lateinit var pstmt: PreparedStatement

    internal fun executeUpdate(sql: String, pss: PreparedStatementSetter) {
        try {
            conn = ConnectionManager.getConnection()
            pstmt = conn.prepareStatement(sql)
            pss.setter(pstmt)

            pstmt.executeUpdate()
        }

        finally {
            pstmt.close()
            conn.close()
        }
    }

    internal fun findById(sql: String, pss: PreparedStatementSetter): User? {
        try {
            conn = ConnectionManager.getConnection()
            pstmt = conn.prepareStatement(sql)
            pss.setter(pstmt)

            val rs = pstmt.executeQuery()

            if (rs.next()) {
                return User(
                    userId = rs.getString("userId"),
                    password = rs.getString("password"),
                    name = rs.getString("name"),
                    email = rs.getString("email")
                )
            }
        }

        finally {
            pstmt.close()
            conn.close()
        }

        return null
    }
}